doc/src/manual/performance-tips.md

```diff
@@ -1507,3 +1507,89 @@ The following examples may help you interpret expressions marked as containing n
+the absolute value of `r`. The two inner functions are called
+"closures". Besides closures, inner functions are also used by the
+language for `do`-blocks and for generator expressions.
+
+This style of code presents performance challenges for the language
+for the following reason. When the parser observes occurrences of
+inner functions, it emits lower-level code in which the inner
+function is pulled out as an object living at global scope. Captured
+variables from the enclosing scope such as `r` must therefore also be moved into
+objects with global scope in order to be accessible to the inner
+function.  This is because the language specifies that `r` in the
+inner scope must be identical to `r` in the outer scope even if the
+outer scope (or another inner function) modifies `r`.
+
+The discussion in the preceding paragraph referred to the "parser", that is, the phase
+of compilation that takes place when the module containing `abmult` is first loaded,
+as opposed to the later phase when it is first invoked. The parser does not "know" that
+`Int` is a fixed type, or that the statement `r = -r` tranforms an `Int` to another `Int`.
+The magic of type inference takes place in the later phase of compilation.
+
+Thus, the parser does not know that `r` has a fixed type (`Int`)
+nor that `r` does not change value once the inner function is created.  Therefore,
+it must create a "box" (a heap allocation) for `r`, and furthermore, the
+box holds an object with an abstract type such as `Any`, which
+requires run-time type dispatch for each occurrence of `r`.  This can be
+verified by applying `@code_warntype` to the above function.  Both the boxing
+and the run-time type dispatch can cause loss of performance.
+
+If captured variables are used in a performance-critical section of the code,
+then the following tips help ensure that their use is performant. First, if
+it is known that a captured variable does not change its type, then this can
+be declared explicitly with a type annotation (on the variable, not the
+right-hand side):
+```julia
+function abmult(r0::Int)
+    r::Int = r0
+    if r >= 0
+        f = x -> x * r
+    else
+        r = -r
+        f = x -> x * r
+    end
+    return f
+end
+```
+The type annotation partially recovers lost performance due to capturing because
+the parser can associate a concrete type to the object in the box.
+Second, if the captured variable does not need to be boxed (because it
+will not be reassigned after the closure is created), this can be indicated
+with `let` blocks as follows.
+```julia
+function abmult3(r::Int)
+    f = if r >= 0
+        let r = r
+            x -> x * r
+        end
+    else
+        let r = -r
+            x -> x * r
+        end
+    end
+    return f
+end
+```
+Each `let` block creates a new variable `r` whose scope is only the
+inner function. The second technique recovers full language performance
+in the presence of captured variables. Note that this is a rapidly
```

Comment:
Maybe mention https://github.com/c42f/FastClosures.jl here, as it automates this process.

Reply:
Is it appropriate to cite non-core (i.e.,  not LinearAlgebra, etc.)  packages in the manual?  I don't see many occurrences of such citations in the manual.  On the other hand, the performance-tips chapter does cite the Lint package, so maybe there is a precedent.

Reply:
ProfileView.jl is also cited.  I'd say it's ok in this instance of a, hopefully, transient wart.